basic integral

BowenFu opened the pull request.

soegaard commented on racket-cas/racket-cas.rkt:
```diff
@@ -1998,6 +2001,183 @@
+  )
+
+(define (definite-integrate u x a b)
+  (define integral (integrate u x))
```
; symbolic computation of the definite integral of u wrt to x with limits a and b.
Is the limits a and b meant to be finite - or is ≠inf.0 allowed?

BowenFu replied:
Only the basic cases are considered in the PR.

soegaard commented on racket-cas/racket-cas.rkt:
```diff
@@ -1998,6 +2001,183 @@
+  ; (check-equal? (integrate '(* (sin (+ (* a (expt x 2)) (* b (expt x 2)))) x) x) ...)
+  ; definite integral
+  (check-equal? (integrate `u x `a `b) '(+ (* -1 a u) (* b u)))
+  )
```
This is already a nice range of expressions that can be integrated.

If you need inspiration, I can recommend looking at: Yacas
https://yacas.readthedocs.io/en/latest/book_of_algorithms/integration.html
The test suite is here:
https://github.com/grzegorzmazur/yacas/blob/master/tests/integrate.yts
And their rule set is here:
https://github.com/grzegorzmazur/yacas/blob/master/scripts/integrate.rep/code.ys

soegaard commented on racket-cas/racket-cas.rkt:
```diff
@@ -2744,6 +2927,10 @@
+                                 (r (Equal v (Exp w)))]
+          [(Equal (Log u v) w)   #:when (and (free-of u x) (free-of w x))
+                                 (r (Equal v (Expt u w)))]
           [(Equal (Expt @e u) s) #:when (> s 0)        (r (Equal u (Ln s)))]
```
Nice catch. Can you make a separate pull request with these improvements of solve?

soegaard commented on racket-cas/racket-cas.rkt:
```diff
@@ -2851,7 +3040,8 @@
-      [(list _ '= x0)   (list x0)]
+      [(list '= y x0) #:when (equal? y x)
+                      (list x0)]
       [_                '()]))
```
Nice catch. Include this in the solve PR.
Is there a test that can catch these problems?

BowenFu commented on racket-cas/racket-cas.rkt:
```diff
@@ -1245,11 +1246,11 @@
+             [(equal? d 0)               (ComplexRealExpt a b c)]
+             [(equal? b 0)               (RealComplexExpt a c d)]
              [else                  (ComplexComplexExpt a b c d)])]
           [(_ _)          `(expt ,u ,v)])])]))
```
I'll create a separate PR for it.

BowenFu commented on racket-cas/racket-cas.rkt:
```diff
@@ -1769,6 +1770,7 @@
     
     [(Acos u) u]    ; xxx only of -1<u<1
+    [(Asin u) (⊖ (⊗ 1/2 @pi) u)]
     [(⊖ u) (Cos u)] ; even function
```
I'll create a separate PR for it.
Actually this simplification is needed to pass some `integrate` test cases.

BowenFu commented:
Thanks for your review. I have now added some doc for the functions. I hope I could have a better name for them.

BowenFu commented on racket-cas/racket-cas.rkt:
```diff
@@ -2868,6 +3230,74 @@
       [(⊖ u)  #f]
       [_      #t])) ; todo: add cases for variables with assumptions
 
+(define (solve-by-inverse-stacks w x)
```
Hi @soegaard , do you have any suggestions on how to share codes between `solve-by-inverse-stacks` and `solve-by-inverse`?

Two differences from `solve-by-inverse`:

1. `solve-by-inverse-stacks` needs to return the list of all intermediate results/eqns.
2. the `let/ec return`.

BowenFu replied:
THe function has been removed.

BowenFu commented:
Hi @soegaard , the MR is almost stable now, can you help review it?

soegaard commented:
Hi,

I just briefly skimmed the PR, but due to the many individual commits, the Github interface makes it hard to get the larger picture.

Can you make a new PR with a single commit - and use `git rebase` to make the commit on top of the current master?

One thing caught my eye though: you have made a matcher Sum to match `(list* '+ us)`.
Although natural I prefer to save the name Sum for a real sum operator like "sum from i=0 to 10 i^2" kind of thing. 

One option is to rename it to, say, HeadPlus (since the head of the form is a plus).
Another is to keep `(list* '+ us)` since this makes it obvious that the pattern matcher doesn't do any rewriting.

I also noted that the integrator can handle quite a large set of expressions - great!

BowenFu commented:
> Hi,
> 
> I just briefly skimmed the PR, but due to the many individual commits, the Github interface makes it hard to get the larger picture.
> 
> Can you make a new PR with a single commit - and use `git rebase` to make the commit on top of the current master?
> 
> One thing caught my eye though: you have made a matcher Sum to match `(list* '+ us)`.
> Although natural I prefer to save the name Sum for a real sum operator like "sum from i=0 to 10 i^2" kind of thing.
> 
> One option is to rename it to, say, HeadPlus (since the head of the form is a plus).
> Another is to keep `(list* '+ us)` since this makes it obvious that the pattern matcher doesn't do any rewriting.
> 
> I also noted that the integrator can handle quite a large set of expressions - great!

Hi @soegaard , the overall changes can be seen in the `File changed` tab. I have rebased my commits now.
Sum / Prod is the orig matcher in the codes. I added two, SumTerms and ProdTerms, now I have renamed them to PlusTerms and TimesTerms.

soegaard commented:
Thanks for rebasing. 

I missed that Sum/Prod was in the original code. Sorry about that.

I am working making racket-cas easier to modify by splitting racket-cas.rkt up in smaller modules.
I hope the change will improve compile time (and maybe make github commits easier to read?).
It's still a work in progress - see the "refactor" branch. It's a bit tricky to avoid circular module dependcies.
I'll keep you posted.

soegaard commented on racket-cas/racket-cas.rkt:
```diff
@@ -828,6 +856,7 @@
     [(Expt (⊕ u v) n) #:when (and (>= n 3) (even? n))
                       (let ([t (e (Expt (⊕ u v) (/ n 2)))])
                         (e (⊗ t t)))]
+    [(Expt (⊕ u v) p-) (Expt (e (Expt (⊕ u v) (- p-))) -1)]
```
What's the intended meaning of the pattern p- ?
Normally p is an integer, but p- also matches -42.5.
Perhaps p- ought to match negative integers only?

    `(check-equal? (math-match -42.5 [p+ #f] [p- #t] [_ #f]) #f)`

BowenFu replied:
You are right. I did not notice the buggy + - before.
We need to fix this.

BowenFu replied:
Is it possible to conjoin the predicates inside find-convention-type?

soegaard commented on racket-cas/racket-cas.rkt:
```diff
@@ -918,11 +948,12 @@
-    [(r  0) +nan.0]
+  (match* (u v)
+    [((num: r)  0) +nan.0]
     [(u  1) u]
```
Why change this?

BowenFu replied:
With math-match*, I'll get an error for the pattern `((TimesTerms (== v) us ...) _)`
`match: incorrect use of ... in pattern in: ...`

BowenFu commented:
> soegaard 3 hours ago  Owner
> What's the intended meaning of the pattern p- ?
> Normally p is an integer, but p- also matches -42.5.
> Perhaps p- ought to match negative integers only?
> 
> `(check-equal? (math-match -42.5 [p+ #f] [p- #t] [_ #f]) #f)`

You are right. I did not notice the buggy + - before.
We need to fix this.
